client/src/worker_registry/mod.rs

```diff
@@ -0,0 +1,275 @@
+//! This module enables the tracking of workers that are associated with a client instance.
+//! This is needed to implement Eager Workflow Start, a latency optimization in which the client,
+//!  after reserving a slot, directly forwards a WFT to a local worker.
+
+use parking_lot::RwLock;
+use slotmap::SlotMap;
+use std::collections::{hash_map::Entry::Vacant, HashMap};
+
+use temporal_sdk_core_protos::temporal::api::workflowservice::v1::PollWorkflowTaskQueueResponse;
+
+slotmap::new_key_type! {
+    /// Registration key for a worker
+    pub struct WorkerKey;
+}
+
+/// This trait is implemented by an object associated with a worker, which provides WFT processing slots.
+#[cfg_attr(test, mockall::automock)]
+pub trait SlotProvider: std::fmt::Debug {
+    /// The namespace for the WFTs that it can process.
+    fn namespace(&self) -> &str;
+    /// The task queue this provider listens to.
+    fn task_queue(&self) -> &str;
+    /// Try to reserve a slot on this worker.
+    fn try_reserve_wft_slot(&self) -> Option<Box<dyn Slot + Send>>;
+}
+
+/// This trait represents a slot reserved for processing a WFT by a worker.
+#[cfg_attr(test, mockall::automock)]
+pub trait Slot {
+    /// Consumes this slot by dispatching a WFT to its worker. This can only be called once.
+    fn schedule_wft(
+        self: Box<Self>,
+        task: PollWorkflowTaskQueueResponse,
+    ) -> Result<(), anyhow::Error>;
+}
+
+/// This trait enables local workers to made themselves visible to a shared client instance.
+/// There can only be one worker registered per namespace+queue_name+client, others will get ignored.
+#[cfg_attr(test, mockall::automock)]
+pub trait WorkerRegistry {
+    /// Register a local worker that can provide WFT processing slots.
+    fn register(&self, provider: Box<dyn SlotProvider + Send + Sync>) -> Option<WorkerKey>;
+    /// Unregister a provider, typically when its worker starts shutdown.
+    fn unregister(&self, id: WorkerKey);
+}
```

Comment:
I'm inclined still to say this trait can just go away and these can be implemented on the `SlotManager` directly - and in general I try to avoid using mocks except when mocking things that require external resources (ex: the existing mocks are pretty much only for clients / pollers), but this is all internal so typically there's not a big reason to mock anything.

The other two traits make a bit more sense to me since they're implemented outside of the client crate.

Reply:
I didn't want "core" to see try_reserve_wft_slot, which can only be called by the client, but I'll remove it...

Reply:
That's what pub(crate) is for

Reply:
Good point, changing the visibility...